### PR TITLE
ref(iroh-net): Remove need for relay info in best_addr

### PR DESCRIPTION
**Pull request by flub**
## Description

This moves metrics for when switches between direct connections and
relayed connections to the NodeState.  This is after all where this
decision is made.  BestAddr only knows about the best UDP address, it
has no business keeping track of relays.

Cleaning this up enables further refactoring of BestAddr and PathState
which need to improve because they depend on each other, but are
hindered by the relay state sneaking into there.

## Breaking Changes

None

## Notes & open questions

These metrics ignore the fact that we have mixed as a possibility so
over-simplify the situation.  This makes the logic to increment them
complex, which in turn makes it no one will really know what they
mean.

I need this to move on with #2546 so don't really want to get into
designing our metrics though.  I believe this way the keep the same
behaviour.

## Change checklist

- [x] Self-review.
- [x] Documentation updates following the [style guide](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#appendix-a-full-conventions-text), if relevant.
- [x] Tests if relevant.
- [x] All breaking changes documented.

### CODE REVIEW

**Review by Arqu — APPROVED** [review]
Nice little state machine you got there :D

File: iroh-net/src/magicsock/node_map/node_state.rs [hunk 1]
```diff
@@ -295,15 +295,39 @@ impl NodeState {
...
+                inc!(MagicsockMetrics, num_relay_conns_added);
+            }
+            if matches!(typ, ConnectionType::Direct(_) | ConnectionType::None)
+                && matches!(
+                    prev_typ,
+                    ConnectionType::Relay(_) | ConnectionType::Mixed(_, _)
+                )
+            {
+                inc!(MagicsockMetrics, num_relay_conns_removed);
+            }
```
Comment (matheus23) [inline iroh-net/src/magicsock/node_map/node_state.rs:330]: I wouldn't mind if this were refactored into a `match (prev_typ, typ)` expression.

Comment (flub) [reply iroh-net/src/magicsock/node_map/node_state.rs]: Be careful what you wish for...  This version is huge and contains a whole bunch of states that are unreachable.  Admittedly the `if` version has to be read backwards, where you think about what conditions are needed for a certain state.  So I guess the big match is less of a head-twister.

Comment (matheus23) [reply iroh-net/src/magicsock/node_map/node_state.rs]: I'd didn't necessarily want it to be an exhaustive match 😅 so yeah I guess I *have* to be careful in what I wish for haha!
Basically, having a last `_ => {}` case would've been fine by me :)

**Pull request merged**